### PR TITLE
perf(lsp): use a stub module in tsc for failed resolutions

### PR DESCRIPTION
**Pull request by nayeemrmn**
Previously when you edited any import statement, the ts server would re-resolve `jsxImportSource` from all modules in the graph. The resolution would return null because it's not a deno_graph-discovered dependency of those modules. Resolutions that return null are frequently retried. Instead make it resolve to a stub module `internal://missing_dependency.d.ts`.

Also resolve `jsxImportSource` up front using the `deno.json` as the referrer.

### CODE REVIEW

**Review by bartlomieju — COMMENTED** [review]
@nayeemrmn is there a test we could add for this change?

File: cli/lsp/tsc.rs [hunk 1]
```diff
@@ -4039,13 +4039,15 @@ fn op_resolve<'s>(
...
-              mt.as_ts_extension().to_string(),
-            )
-          })
+        .map(|o| match o {
+          Some((s, mt)) => Some((
+            state.specifier_map.denormalize(&s),
+            mt.as_ts_extension().to_string(),
+          )),
+          None => Some((
+            "internal://missing_dependency.d.ts".to_string(),
```
Comment (bartlomieju) [inline cli/lsp/tsc.rs:4048]: We will not have 6 occurrences of this string. Maybe make it a static somewhere?

File: cli/lsp/tsc.rs [hunk 2]
```diff
@@ -4026,11 +4027,11 @@ fn op_load<'s>(
 }
 
 #[op2]
+#[serde]
 fn op_resolve<'s>(
-  scope: &'s mut v8::HandleScope,
   state: &mut OpState,
   #[serde] args: ResolveArgs,
-) -> Result<v8::Local<'s, v8::Value>, AnyError> {
+) -> Result<Vec<Option<(String, String)>>, AnyError> {
```
Comment (nayeemrmn) [inline cli/lsp/tsc.rs:4034]: We made this return a v8 local to include the serialization in the performance mark, I reverted that to make it more easily unit testable with `op_resolve::call()`. Not sure it was particularly helpful anyway.

File: cli/tsc/99_main_compiler.js [hunk 3]
```diff
@@ -725,10 +725,6 @@ delete Object.prototype.__proto__;
           if (item) {
             isCjsCache.add(item);
             const [resolvedFileName, extension] = item;
-            if (resolvedFileName.startsWith("node:")) {
-              // probably means the user doesn't have @types/node, so resolve to undefined
-              return undefined;
-            }
```
Comment (bartlomieju) [inline cli/tsc/99_main_compiler.js:731]: And this one?

File: cli/lsp/tsc.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/lsp/tsc.rs]: That's fine 👍

File: cli/tsc/mod.rs [hunk 4]
```diff
@@ -575,14 +577,12 @@ fn op_resolve(
...
   };
   for specifier in args.specifiers {
-    if let Some(module_name) = specifier.strip_prefix("node:") {
-      if deno_node::is_builtin_node_module(module_name) {
-        // return itself for node: specifiers because during type checking
-        // we resolve to the ambient modules in the @types/node package
-        // rather than deno_std/node
-        resolved.push((specifier, MediaType::Dts.to_string()));
-        continue;
-      }
```
[the remark this answers is not in the document]
Comment (nayeemrmn) [reply cli/tsc/mod.rs:585]: `node:` specifiers were being forced to `undefined` on the JS side, so I'm not sure what this was doing probably poorly aged code. Anyway I moved that to rust to factor out `MISSING_DEPENDENCY_SPECIFIER`.

File: cli/tsc/99_main_compiler.js (hunk 3 above)
Comment (nayeemrmn) [reply cli/tsc/99_main_compiler.js:731]: Moved to rust in `op_resolve()` implementations.

**Comment by bartlomieju** [comment]
@dsherret can you take a look and double check? It looks good to me

File: cli/tsc/mod.rs [hunk 5]
```diff
@@ -444,6 +443,9 @@ pub fn as_ts_script_kind(media_type: MediaType) -> i32 {
   }
 }
 
+pub const MISSING_DEPENDENCY_SPECIFIER: &str =
+  "internal:///missing_dependency.d.ts";
```
Comment (dsherret) [inline cli/tsc/mod.rs:447]: This is the source of a lot of confusion in Deno when dependencies can't resolve, but it ends up as any.

File: cli/tsc/mod.rs [hunk 6]
```diff
@@ -575,14 +577,12 @@ fn op_resolve(
...
-        resolved.push((specifier, MediaType::Dts.to_string()));
-        continue;
-      }
+    if specifier.starts_with("node:") {
+      resolved.push((
+        MISSING_DEPENDENCY_SPECIFIER.to_string(),
+        MediaType::Dts.to_string(),
+      ));
+      continue;
     }
```
Comment (dsherret) [inline cli/tsc/mod.rs:586]: Will this cause invalid node specifiers to not surface an error?

Comment (nayeemrmn) [reply cli/tsc/mod.rs:586]: Still works, those errors aren't surfaced from returning `None` here.

**Review by dsherret — APPROVED** [review]
LGTM

**Pull request merged**